raft/src/main/java/org/apache/kafka/raft/RequestManager.java

```diff
@@ -17,108 +17,196 @@
+                res = Optional.empty();
                 break;
             }
         }
+
         return res;
     }
 
-    public long backoffBeforeAvailableVoter(long currentTimeMs) {
-        long minBackoffMs = Long.MAX_VALUE;
-        for (Integer voterId : voters) {
-            ConnectionState connection = connections.get(voterId);
-            if (connection.isReady(currentTimeMs)) {
+    public long backoffBeforeAvailableBootstrapServer(long currentTimeMs) {
+        long minBackoffMs = Math.max(retryBackoffMs, requestTimeoutMs);
+        for (Node node : bootstrapServers) {
+            if (isReady(node, currentTimeMs)) {
                 return 0L;
-            } else if (connection.isBackingOff(currentTimeMs)) {
-                minBackoffMs = Math.min(minBackoffMs, connection.remainingBackoffMs(currentTimeMs));
+            } else if (isBackingOff(node, currentTimeMs)) {
+                minBackoffMs = Math.min(minBackoffMs, remainingBackoffMs(node, currentTimeMs));
             } else {
-                minBackoffMs = Math.min(minBackoffMs, connection.remainingRequestTimeMs(currentTimeMs));
+                minBackoffMs = Math.min(minBackoffMs, remainingRequestTimeMs(node, currentTimeMs));
             }
         }
+
         return minBackoffMs;
     }
 
+    public boolean hasRequestTimedOut(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return false;
+        }
+
+        return state.hasRequestTimedOut(timeMs);
+    }
+
+    public boolean isReady(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return true;
+        }
+
+        boolean ready = state.isReady(timeMs);
+        if (ready) {
+            reset(node);
+        }
+
+        return ready;
+    }
+
+    public boolean isBackingOff(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return false;
+        }
+
+        return state.isBackingOff(timeMs);
+    }
+
+    public long remainingRequestTimeMs(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return 0;
+        }
+
+        return state.remainingRequestTimeMs(timeMs);
+    }
+
+    public long remainingBackoffMs(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return 0;
+        }
+
+        return  state.remainingBackoffMs(timeMs);
+    }
+
+    public boolean isResponseExpected(Node node, long correlationId) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return false;
+        }
+
+        return state.isResponseExpected(correlationId);
+    }
+
+    public void onResponseReceived(Node node, long correlationId) {
```

Comment:
I wonder if we could consolidate `onResponseReceived` and `onResponseError`. Perhaps something like `notifyResponseResult` or something like that.

Reply:
Sounds good. Went with `onResponseResult(Node, long, boolean, long)`.